odoo/models.py

```diff
@@ -6709,6 +6721,53 @@ def flush_recordset(self, fnames=None):
         if self.env.cache.has_dirty_fields(self, fields_):
             self._flush(fnames)
 
+    def _ensure_exists(self):
+        if not self:
+            return
+        precommit = self.env.cr.precommit
+        if 'ensure_exists' not in precommit.data:
+            precommit.data['ensure_exists'] = defaultdict(set)
+            precommit.add(self._ensure_exists_callback)
+        precommit.data['ensure_exists'][self._name].update(self._ids)
+
+    def _ensure_exists_callback(self):
+        """ enusre records' existence in the current and all concurrent transactions """
+        data = self.env.cr.precommit.data.get('ensure_exists')
+        if not data:
+            return
+        for model_name, ids in data.items():
+            if not ids:
+                continue
+            model = self.env[model_name]
+            # Lock referenced records by performing a dummy update to prevent concurrent deletion
+            self.env.cr.execute(SQL("""
+                UPDATE %(table)s
+                SET id = id
+                WHERE id IN %(ids)s
+                """,
+                table = SQL.identifier(model._table),
+                ids = tuple(ids),
+            ))
+            if self.env.cr.rowcount == len(ids):
+                continue
+            missing_ids = self.env.execute_query(SQL(
+                """
+                WITH ref_ids AS (
+                    SELECT unnest(%(ids)s) AS id
+                )
+                SELECT array_agg(r.id)
+                FROM ref_ids r
+                WHERE NOT EXISTS (
+                    SELECT 1 
+                    FROM %(table)s t
+                    WHERE t.id = r.id
+                )
+                """,
+                ids = list(ids),
+                table = SQL.identifier(model._table),
+            ))[0][0]
```

Comment:
Wouldn't this be simpler and as effective?
```suggestion
            records = model.browse(ids)
            missing = records - records.exists()
```